Use SVG external content for social-logos.

sgomes opened the pull request:
Reimplement social-logos using an SVG external content approach, instead of SVG transpiled into React components. This helps keep images as images (which is good for performance), while maintaining styleability. Old browsers are supported through the already-included `svg4everyone` polyfill.

See #32364 for a similar change, which added Material Icons with the same approach.

This PR moves ~15KB from JS to SVG in several sections and async chunks.

#### Changes proposed in this Pull Request

* Bump `social-logos` package version to `2.1.0`, which includes the SVG spritemap.
* Implement `components/social-logo` using SVG external content approach.
* Rewrite all uses of `SocialLogo` to use `components/social-logo` instead of the `social-logos` package.
* Replace usage of example in `social-logos` package with a new example in `components/social-logo/docs`.

#### Testing instructions

* Test as many places where social logos are used as possible. Ensure that logos look identical to before this change.

sgomes committed:
Use SVG external content for social-logos.

Reimplement social-logos using an SVG external content approach,
instead of SVG transpiled into React components. This helps keep
images as images (which is good for performance), while maintaining
styleability. Old browsers are supported through the already-included
`svg4everyone` polyfill.

ockham commented:
Devdocs still pull in example content from the package: https://github.com/Automattic/wp-calypso/blob/48b121e7ce115566888e350b9deae4a47e6d1a0f/client/devdocs/design/index.jsx#L104

(using this alias: https://github.com/Automattic/wp-calypso/blob/48b121e7ce115566888e350b9deae4a47e6d1a0f/webpack.config.js#L252)

sgomes commented:
> Devdocs still pull in example content from the package:
> 
> https://github.com/Automattic/wp-calypso/blob/48b121e7ce115566888e350b9deae4a47e6d1a0f/client/devdocs/design/index.jsx#L104
> 
> (using this alias:
> 
> https://github.com/Automattic/wp-calypso/blob/48b121e7ce115566888e350b9deae4a47e6d1a0f/webpack.config.js#L252
> 
> )

Yes, I intentionally left it that way, since I assume not a lot of regular users use devdocs, and since the example actually lives outside this repo, inside `social-logos`.

We could of course reimplement the example in Calypso using the new approach. Do you think that would be best?

sgomes commented:
Currently investigating test failure. It seems the transformation we use in tests to enable compatibility with `file-loader` isn't working here, for some reason.

sgomes committed:
Apply file-loader jest transform to social-logos.

sgomes commented:
@jsnajdr 

> Currently investigating test failure. It seems the transformation we use in tests to enable compatibility with `file-loader` isn't working here, for some reason.

The latest commit should fix this. `social-logos` was being excluded from the jest transform that enables `file-loader` compatibility.

ockham commented:
> > Devdocs still pull in example content from the package:
> > https://github.com/Automattic/wp-calypso/blob/48b121e7ce115566888e350b9deae4a47e6d1a0f/client/devdocs/design/index.jsx#L104
> > 
> > (using this alias:
> > https://github.com/Automattic/wp-calypso/blob/48b121e7ce115566888e350b9deae4a47e6d1a0f/webpack.config.js#L252
> > 
> > )
> 
> @ockham: Yes, I intentionally left it that way, since I assume not a lot of regular users use devdocs, and since the example actually lives outside this repo, inside `social-logos`.
> 
> We could of course reimplement the example in Calypso using the new approach. Do you think that would be best?

Maybe -- but then again, it's really not that critical :+1: 

I was mostly thinking about a future `social-logos` package that drops the JS wrapper altogether and only ships the SVG. Not sure if it would/should then still ship a pre-built example, or if that should be Calypso's/DevDocs' concern :thinking:

Arguably, it should be consistent with Gridicons, which currently do have their own (Calypso-independent) [website](http://automattic.github.io/gridicons/) :thinking:

sgomes commented:
> Arguably, it should be consistent with Gridicons, which currently do have their own (Calypso-independent) [website](http://automattic.github.io/gridicons/) 🤔

I suppose at that point you could consider it two separate examples: usage of the default SVG-in-JS icons shipped with the packages, and usage of the Calypso wrapper for the icons. And if we were ever to ship `social-logos` as just the SVG, we could reevaluate.

I'll implement an example in Calypso using the new approach, then 🙂

sgomes committed:
Add SocialLogo example for devdocs.

We no longer make use of the example included in the
social-logos package, instead implementing our own.

sgomes commented:
@ockham Removed the dependency on the `social-logos` package example and added a new example in Calypso.

sgomes commented on client/devdocs/design/component-examples.js:
```diff
@@ -1,93 +1,94 @@
-export Wizard from 'components/wizard/docs/example';
-export WizardProgressBar from 'components/wizard-progress-bar/docs/example';
-export WpcomColophon from 'components/wpcom-colophon/docs/example';
+export { default as ActionCard } from 'components/action-card/docs/example';
```
Note: these changes were made to appease ESLint, which doesn't handle `eslint-disable` correctly for parsing errors, such as the one the previous syntax caused.

sgomes replied:
Outdated.

sgomes commented on client/devdocs/design/playground-scope.js:
```diff
@@ -1,123 +1,123 @@
  */
-export Gridicon from 'gridicons';
-export SocialLogo from 'social-logos';
+export { default as Gridicon } from 'gridicons';
```
Note: these changes were made to appease ESLint, which doesn't handle `eslint-disable` correctly for parsing errors, such as the one the previous syntax caused.

sgomes replied:
Outdated.

sgomes committed:
Fix onClick handler in SharingButtonsPreviewButton

sgomes commented on client/my-sites/marketing/buttons/preview-button.jsx:
```diff
@@ -60,14 +54,16 @@ export default createReactClass( {
+	/* eslint-enable wpcalypso/jsx-classname-namespace */
 
-	onClick: function() {
+	onClick() {
```
Good catch! Fixed.

jsnajdr commented:
#32761 should improve the TypeScript linting situation. `export` extensions in JS files should stop breaking things, and there is only one `no-unused-vars` rule.

sgomes commented:
> #32761 should improve the TypeScript linting situation. `export` extensions in JS files should stop breaking things, and there is only one `no-unused-vars` rule.

I'll put this PR on hold until #32761 gets merged. Once it does, I'll remove all the ESLint-related changes. Feel free to either review now with that in mind, or wait until then to review!

sgomes commented:
This PR has now been rebased on top of #32761 and is ready for a final review!

sgomes committed:
Fix linting error in devdocs design.

blowery reviewed (approved):
Looks good to me! Nice work.

sgomes commented:
Thanks for the review, @blowery! 👍 I'll wait for @jsnajdr's ok too before merging.
